buildkite/package_uploads.go

```diff
@@ -0,0 +1,216 @@
+	RegistrySlug     string `json:"-"`
+
+	URI  string                     `json:"uri"`
+	Form PackagePresignedUploadForm `json:"form"`
+}
+
+type PackagePresignedUploadForm struct {
+	FileInput string            `json:"file_input"`
+	Method    string            `json:"method"`
+	URL       string            `json:"url"`
+	Data      map[string]string `json:"data"`
+}
+
+// RequestPresignedUpload requests a presigned upload URL for a Buildkite package from the buildkite API
+func (ps *PackagesService) RequestPresignedUpload(ctx context.Context, organizationSlug, registrySlug string) (PackagePresignedUpload, *Response, error) {
+	url := fmt.Sprintf("v2/packages/organizations/%s/registries/%s/packages/upload", organizationSlug, registrySlug)
+	req, err := ps.client.NewRequest(ctx, "POST", url, nil)
+	if err != nil {
+		return PackagePresignedUpload{}, nil, fmt.Errorf("creating POST presigned upload request: %v", err)
+	}
+
+	var p PackagePresignedUpload
+	resp, err := ps.client.Do(req, &p)
+	if err != nil {
+		return PackagePresignedUpload{}, resp, fmt.Errorf("executing POST presigned upload request: %v", err)
+	}
+
+	p.OrganizationSlug = organizationSlug
+	p.RegistrySlug = registrySlug
+
+	return p, resp, err
+}
+
+// Perform performs uploads the package file referred to by `file` to the presigned upload URL.
+// It does not create the package in the registry, only uploads the file to S3. The returned string is the URL of the
+// uploaded file in S3, which can then be passed to [Finalize] to create the package in the registry.
+func (ppu PackagePresignedUpload) Perform(ctx context.Context, ps *PackagesService, file *os.File) (string, error) {
+	if _, ok := ppu.Form.Data["key"]; !ok {
+		return "", fmt.Errorf("missing 'key' in presigned upload form data")
+	}
+
+	baseFilePath := filepath.Base(file.Name())
+
+	s := bkmultipart.NewStreamer()
+	err := s.WriteFields(ppu.Form.Data)
+	if err != nil {
+		return "", fmt.Errorf("writing form fields: %v", err)
+	}
+
+	err = s.WriteFile(ppu.Form.FileInput, file, baseFilePath)
+	if err != nil {
+		return "", fmt.Errorf("writing form file: %v", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, ppu.Form.Method, ppu.Form.URL, s.Reader())
+	if err != nil {
+		return "", fmt.Errorf("creating %s request: %v", ppu.Form.Method, err)
+	}
+
+	req.Header.Set("Content-Type", s.ContentType)
+
+	// Don't set the Content-Length header here, you fool, you absolute buffoon
+	// When passed an io.Reader, http.NewRequestWithContext will not set the Content-Length header, and will instead
+	// stream the request body. This _would_ be exactly what we want, except that S3 uploads infuriatingly require a
+	// Content-Length header. So we have to calculate the length of the request body ourselves and set it manually on the
+	// request. Adding:
+	// 	req.Header.Set("Content-Length", fmt.Sprintf("%d", s.Len()))
+	// is not sufficient, as the Content-Length header is stripped by the http client when the request body is an io.Reader.
```

Comment:
i'm not particularly tied to this comment, it was just something i learned along the way. happy to get rid of it or keep it, if anyone has opinions
```suggestion
```